### DISCUSSION TITLE
Function acts volatile even though it is set to false

### DISCUSSION THREAD

**Discussion by simemese** [Excel-DNA/ExcelDna#447] [Q&A] [answered]
We have functions in this structure that act volatile even though they are set to not be volatile (e.g. recalculates if a row is added). Does anyone have any idea what are doing wrong?

 [ExcelFunction(Description = "Grabs available curves. Please specify the quarter in this format: 2020Q4.", IsVolatile = false)]
        public static string fl_DownloadCurves(string quarter, string filePath)
        {
            try
            {
                if (String.IsNullOrEmpty(filePath))
                {
                    filePath = @"B:\Curves" + quarter + ".xls";
                }

                Curves curves = new Curves(filePath);
                return dataStore.AddObjectAndReturnKey(curves);
            }
            catch (Exception ex)
            {

                return ex.Message;
            }
        }

**Comment by govert** [comment, marked as answer]
You are not doing anything wrong  - you are seeing normal behaviour for any function in Excel, unrelated to the 'Volatile' concept.
Excel recalculates the sheet when making structural changes like adding or deleting a row or column.

Compare the behaviour of a VBA function like this:
```vb
Function Increment()
    Static i As Integer
    
    i = i + 1
    Increment = i 
End Function
```

With a volatile function like `=RAND()`, putting both into a sheet.

You'll see that only the volatile function `=RAND()` calculates when you edit some unrelated cell on the sheet, but both calculate if you add a row.

**Reply by simemese** [reply]
Thanks! Ended up caching the results and returning if they already exist.

https://excel-dna.net/2013/04/02/caching-and-asynchronous-excel-udfs/
